src/emmeans.jl

```diff
@@ -0,0 +1,187 @@
+    discussed in [the documentation for the R package `emmeans`](https://cran.r-project.org/web/packages/emmeans/vignettes/transformations.html).
+"""
+function empairs(model::RegressionModel; eff_col=nothing, err_col=:err,
+                 invlink=identity, levels=Dict(), dof=nothing, padjust=identity)
+    eff_col = something(eff_col, _responsename(model))
+    em = emmeans(model; eff_col, err_col, invlink, levels, dof)
+    return empairs(em; eff_col, err_col, padjust)
+end
+
+function empairs(df::AbstractDataFrame; eff_col, err_col=:err, padjust=identity)
+    # need to enforce that we're all the same type
+    # (mixing string and symbol is an issue with Not
+    #  and a few other things below)
+    eff_col = string(eff_col)
+    err_col = string(err_col)
+    stats_cols = [eff_col, err_col]
+    "dof" in names(df) && push!(stats_cols, "dof")
+
+    pairs = combinations(eachrow(df), 2)
+    # TODO make this more efficient in allocations
+    result_df = mapreduce(vcat, pairs) do (df1, df2)
+        result = Dict{String,Union{String,Number}}()
+
+        for col in names(df1, Not(stats_cols))
+            result[col] = if df1[col] == df2[col]
+                df1[col]
+            else
+                string(df1[col], " > ", df2[col])
+            end
+        end
+        result[eff_col] = df1[eff_col] - df2[eff_col]
+        result[err_col] = pooled_sem(df1[err_col], df2[err_col])
+        if "dof" in names(df)
+            df1["dof"] != df2["dof"] &&
+                throw(ArgumentError("Pooled dof not suported: $(df1) and $(df2)"))
+            result["dof"] = df1["dof"]
+        end
+        return DataFrame(result)
+    end
+
+    # XXX post process to remove extra rows
+    # combinations() will generate pairs of rows that contrast in some but
+    # not all columns, but we don't want those rows for columns that actually
+    # have available contrasts, so we drop them here.
+    # this is inefficient because we're generating values that we throw away
+    # but it's a lot easier than trying to only generate the correct contrasts
+    cols_with_contrasts = filter(col -> length(unique(result_df[!, col])) > 1,
+                                 names(df, Not(stats_cols)))
+    length(cols_with_contrasts) > 1 &&
+        subset!(result_df,
+                (col => ByRow(contains(" > ")) for col in cols_with_contrasts)...)
+
+    cols = vcat(names(df, Not(stats_cols)), stats_cols)
+    select!(result_df, cols)
+    if "dof" in stats_cols
+        transform!(result_df, [eff_col, err_col] => ByRow(/) => "t")
+        transform!(result_df,
+                   [:dof, :t] => ByRow() do dof, t
+                       p = 2 * cdf(TDist(dof), -abs(t))
+                       return p
+                   end => "Pr(>|t|)")
+        transform!(result_df, "Pr(>|t|)" => padjust => "Pr(>|t|)")
+    elseif padjust !== identity
+        @warn "padjust specified, but there are no p-values to adjust."
```

Comment:
It's pretty atypical in Julia to emit warnings like this. I kinda feel like this should be an `ArgumentError`.

Reply:
It's not really an error though. There's just nothing to do.

Reply:
I'd just ignore it and document that it's ignored in that case